students/grant_dowell/Lesson_04/mailroom_part2.py

```diff
@@ -0,0 +1,121 @@
+               " ~ The Treasurer"
+    donor_info = {'name':name, 'last_donation':db[name][-1]}
+    letter = template.format(**donor_info)
+    return letter
+
+def thanks():
+    """ Generates a Thank You message for a new donation, and adds donation
+        to database """
+
+    thank_you = ""
+    vld_name = False
+    while vld_name is False:
+        name = input("Please enter a Full Name or 'List': ")
+        if name.lower() == 'list':
+            for name in db.keys():
+                print(name)
+        elif name.lower() == 'quit':
+            break
+        else:
+            vld_name = True
+            for _ in db:    # Check for name in db
+                if name.lower == name.lower:
+                    break
+            else:   # If name not found in db, add it
+                db[name] = []
+    else:
+        print(name)
+        donation = input("Please enter a donation amount: ")
+        if donation.lower() != 'quit':  #Check if user is attempting to quit
+            db[name].append(float(donation))
+            thank_you = generate_letter(name)
+        print("\n" + thank_you + "\n")
+    return db, thank_you
+
+def report():
+    """Generates report of donors and donation statistics, sorted by total
+       given."""
+
+    print("Donor Name            | Total Given | Num Gifts | Average Gift")
+    print("--------------------------------------------------------------")
+    for key, value in sorted(db.items(), key=sortby_amt, reverse=True):
+        print("{:<21}  ${:11.2f}   {:9d}  ${:12.2f}".format(key,
+              sum(value), len(value), (sum(value)/len(value))))
+    print('\n\n')
+
+def log_all_letters():
+    """ Generates a letter for every person in the global database and saves
+        it as a separate file."""
+    for name in db:
+        file_name = name + '.txt'
+        with open(file_name, 'w') as outfile:
+            outfile.write(generate_letter(name))
+
+def quitter():
+    print('Goodbye!')
+
+
+
+# Init the donor database
+db = {"William Gates, III": [653772.32, 12.17],
+      "Jeff Bezos": [877.33],
+      "Paul Allen": [663.23, 43.87, 1.32],
+      "Mark Zuckerberg": [1663.23, 4300.87, 10432.0],
+      "John Doe": [1.00, 2.00, 3.00, 4.00]}
+
+menu_quit = False
+
+if __name__ == '__main__':
+#    while True:
+#        print('Select an Operation:')
+#        print("  1) Send Thank You")
+#        print("  2) Create a Report")
+#        print("  3) Quit\n")
+#        cmd = input('> ')
+#        if cmd == '3' or cmd.lower() == 'quit':
+#            print('Goodbye!')
+#            break
+#        elif cmd == '1' or cmd.lower() == 'send thank you':
+#            donor_db, note = thanks(donor_db)
+#            print(note)
+#            print('\n')
+#        elif cmd == '2' or cmd.lower() == 'create a report':
+#            report(donor_db)
+#        else:
+#            continue
+
+    menu = {"1":thanks, "2": report, "3":log_all_letters, "4": quitter}
+    cmd = None
+    while cmd != '4':
```

Comment:
this is not necessary, when user selects option "4" your `quitter()` function should call `sys.exit()` to exit the script.

so in this case your `while True` loop would just terminate

Reply:
I was having some trouble with this giving a Traceback originally. I did a bit more digging though, and I think it was a result of running with the '-i' option. Fixed.